### PR TITLE
chore(internal): offer readonly api in core

### PR DESCRIPTION
**Pull request by emmettbutler**
This change adds `core.set_safe`, which throws an error when the given key already exists on the current context. This function is intended for use in cases where code in `ddtrace.contrib` needs to set a value on an existing context. Product code (`ddtrace.apm` and `ddtrace.appsec`, for example) can still use `core.set_item`.

Read-only semantics for at least one side of the contrib/product shared state setup goes a long way toward avoiding this code turning into a hairball.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/internal/core.py [hunk 1]
```diff
@@ -144,7 +144,6 @@ def addParent(self, context):
         if self.identifier == ROOT_CONTEXT_ID:
             raise ValueError("Cannot add parent to root context")
         self._parents.append(context)
-        self._data.update(context._data)
```
Comment (mabdinur) [inline ddtrace/internal/core.py:147]: Was this change intentional?

**Review by mabdinur — DISMISSED** [review]

File: ddtrace/internal/core.py (hunk 1 above)
Comment (emmettbutler) [reply ddtrace/internal/core.py:147]: Yes. It removes one of the ways that contribs can mutate the contents of an `ExecutionContext`.

**Review by mabdinur — APPROVED** [review]
